core/server/master/src/main/java/alluxio/master/file/loadmanager/LoadManager.java

```diff
@@ -87,393 +83,276 @@ public final class LoadManager {
-     * @param status task status
-     */
-    public void addUfsLoadStatus(String ufs, TaskStatus status) {
-      mUfsLoadTaskStatusMap.put(ufs, status);
-    }
-
-    /**
-     * Record failed block information.
-     * @param blockId block id
-     * @param error error code and messages
-     */
-    public void addFailedBlock(long blockId, String error) {
-      if (!mFailedBlockMap.containsKey(blockId)) {
-        mFailedBlockMap.put(blockId, Sets.newHashSet());
+  // Returns false if the whole task fails
+  private boolean processTask(Load load, LoadTask loadTask)
+  {
+    Preconditions.checkArgument(loadTask.isDone(), "task is not done");
+    try {
+      LoadResponse response = loadTask.getResponse();
+      if (response.getStatus() != TaskStatus.SUCCESS) {
+        for (BlockStatus status : response.getBlockStatusList()) {
+          if (!status.getRetryable() || !load.addBlockToRetry(status.getBlock())) {
+            load.addBlockFailure(
+                status.getBlock(), Status.fromCodeValue(status.getCode()));
+          }
+        }
       }
-      mFailedBlockMap.get(blockId).add(error);
+      return response.getStatus() != TaskStatus.FAILURE;
     }
-
-    /**
-     * Set worker load finish status.
-     * @param address worker address
-     * @param s worker load status
-     */
-    public void setWorkerLoadFinished(WorkerNetAddress address, Boolean s) {
-      mWorkerLoadFinished.put(address, s);
+    catch (CancellationException | ExecutionException e) {
+      loadTask.getBlocks().forEach(load::addBlockToRetry);
+      return false;
     }
-  }
-
-  static class BlockBuffer {
-    static final int BATCH_SIZE = 50;
-    private final Lock mLock = new ReentrantLock();
-    private final List<FileInfo> mFileInfoList;
-    private final List<Block> mBufferedFileBlocks = Lists.newArrayList();
-    private List<Block> mRetriedFileBlocks = Lists.newArrayList();
-    private List<Block> mFailedFileBlocks = Lists.newArrayList();
-    private final Iterator<FileInfo> mFileInfoIterator;
-    private Iterator<Long> mBlockIdIterator;
-    private FileInfo mFileInfo = null;
-
-    /**
-     * Constructor to create a BlockBuffer.
-     * @param fileInfoList FileInfo list
-     */
-    private BlockBuffer(List<FileInfo> fileInfoList) {
-      mFileInfoList = fileInfoList;
-      mFileInfoIterator = mFileInfoList.iterator();
+    catch (InterruptedException e) {
+      loadTask.getBlocks().forEach(load::addBlockToRetry);
+      Thread.currentThread().interrupt();
+      // We don't count InterruptedException as task failure
+      return true;
     }
+  }
 
-    /**
-     * Create a BlockBuffer.
-     * @param fileInfoList fileInfoList param to use
-     * @return a BlockBuffer instance
-     */
-    public static  BlockBuffer create(List<FileInfo> fileInfoList) {
-      return new BlockBuffer(fileInfoList);
+  private Optional<LoadTask> scheduleBatch(
+      Load load,
+      CloseableResource<BlockWorkerClient> workerClient,
```

Comment:
any particular reason to pass in a CloseableResource of worker here?

Reply:
That's what's returned from `mContext.acquireBlockWorkerClient`. I have no idea why that's a `CloseableResource`.